src/extra/components/pulses.py

```diff
@@ -0,0 +1,422 @@
+            type(self).__name__, self._sase, self._source.source)
+
+    @classmethod
+    def _identify_sase(cls, data):
+        """Try to identify which SASE this data belongs to."""
+
+        sases = {sase
+                 for src in data.all_sources
+                 for sase, topics in cls._sase_topics.items()
+                 if src[:src.find('_')] in topics}
+
+        if len(sases) == 1:
+            return sases.pop()
+        elif sases == {1, 3}:
+            # SA3 data often contains one or more SA1 sources
+            # from shared upstream components.
+            return 3
+        else:
+            raise ValueError('sources from ambiguous sase branches {} found, '
+                             'please pass explicit sase'.format(
+                                ', '.join(sases)))
+
+    def _get_pulse_mask(self, table):
+        return is_sase(table, sase=self._sase)
+
+    @property
+    def sase(self) -> int:
+        """Used SASE beamline."""
+        return self._sase
+
+
+class PulsePatternPPL(_PulsePattern):
+    """An interface to optical laser pulse patterns.
+
+    The pump-probe lasers (PPL) are optical lasers commonly used in
+    conjunction with X-ray pulses for pump-probe experiments. Their
+    pulse pattern can be independent
+
+    Args:
+        data (DataCollection): Data to access timing data from.
+        timeserver (str, optional): Source name of a timeserver, only
+            needed if the data includes more than one timeserver or it
+            could not be detected automatically.
+        ppl_seed (euxfel_bunch_pattern.PPL_BITS, optional): PPL seed
+            to interpret pulses of, only needed if the data includes
+            sources from more than one instrument or it could not be
+            detected automatically.
+    """
+
+    _instrument_ppl_seeds = {
+        'FXE': PPL_BITS.LP_FXE,
+        'SPB': PPL_BITS.LP_SPB,
+        'MID': PPL_BITS.LP_SASE2,
+        'HED': PPL_BITS.LP_SASE2,
+        'SCS': PPL_BITS.LP_SCS,
+        'SQS': PPL_BITS.LP_SQS
+    }
+
+    def __init__(self, data, timeserver=None, ppl_seed=None):
+        super().__init__(data, timeserver)
+
+        if ppl_seed is None:
+            ppl_seed = self._identify_ppl_seed(data)
+
+        self._ppl_seed = ppl_seed
+
+    def __repr__(self):
+        return "<{} for {} using timeserver={}>".format(
+            type(self).__name__, self._ppl_seed.name, self._source.source)
+
+    @classmethod
+    def _identify_ppl_seed(cls, data):
+        """Try to identify which PPL seed this data belongs to."""
+
+        instruments = {instrument for src in data.all_sources
+                       if ((instrument := src[:src.find('_')])
+                           in cls._instrument_ppl_seeds.keys())}
+
+        if len(instruments) == 1:
+            return cls._instrument_ppl_seeds[instruments.pop()]
+        elif len(instruments) > 1:
+            raise ValueError('sources from ambiguous instruments {} found,
```

Comment:
This line is missing a trailing quote:

```suggestion
            raise ValueError('sources from ambiguous instruments {} found, '
```

Reply:
That happens when you add things after running tests :see_no_evil: